### PR TITLE
[web] - Fix autofill group input ordering 

### PR DESCRIPTION
**Pull request by htoor3**
Ordering of input elements inside of the DOM tree for autofill groups does not reflect the order of the form rendered on screen.  This is causing some issues with password managers and autofill, specifically Bitwarden.

We are currently always appending the currently focused input element to the end of the form. 

This leads to a tree that appears out of order:
Screenshot 2023-05-23 at 2 57 37 PM

This fix is tracking the position of where the focused input node should be inserted and inserting it there, rather than always at the end of the form.  Once the tree is ordered correctly, Bitwarden's autofill logic works in Flutter forms. 

Tree order after fix:
Screenshot 2023-05-23 at 6 01 05 PM

Fixes https://github.com/flutter/flutter/issues/61301

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 1]
```diff
@@ -153,6 +154,7 @@ class EngineAutofillForm {
 
   final Map<String, AutofillInfo>? items;
 
+  final DomHTMLElement? nodeToInsertBefore;
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:157]: As discussed offline, let's change this to a more descriptive name.

**Review by mdebbar — COMMENTED** [review]
LGTM. Thanks for fixing this!

Once tests are added, please re-ping me for re-review.

File: lib/web_ui/test/engine/text_editing_test.dart [hunk 2]
```diff
@@ -2176,6 +2176,42 @@ Future<void> testMain() async {
...
+              createAutofillInfo('email', 'field1'), fields)!;
+
+      expect(autofillForm.elements, hasLength(2));
+
+      List<DomHTMLInputElement> formChildNodes =
+          autofillForm.formElement.childNodes.toList() as List<DomHTMLInputElement>;
+
+      // Only username, password, submit nodes are created
+      expect(formChildNodes, hasLength(3));
+      expect(formChildNodes[0].name, 'username');
```
Comment (mdebbar) [inline lib/web_ui/test/engine/text_editing_test.dart:2200]: Let's check all children.
```suggestion
      expect(formChildNodes, hasLength(3));
      expect(formChildNodes[0].name, 'username');
      expect(formChildNodes[1].type, 'password');
      expect(formChildNodes[2].type, 'submit');
```

File: lib/web_ui/test/engine/text_editing_test.dart [hunk 3]
```diff
@@ -2176,6 +2176,42 @@ Future<void> testMain() async {
...
+
+      final DomHTMLInputElement testInputElement = createDomHTMLInputElement();
+      testInputElement.name = 'email';
+      autofillForm.placeForm(testInputElement);
+
+      formChildNodes = autofillForm.formElement.childNodes.toList()
+          as List<DomHTMLInputElement>;
+      // email node should be placed before username
+      expect(formChildNodes[0].name, 'email');
+      expect(formChildNodes[1].name, 'username');
```
Comment (mdebbar) [inline lib/web_ui/test/engine/text_editing_test.dart:2212]:
```suggestion
      formChildNodes = autofillForm.formElement.childNodes.toList()
          as List<DomHTMLInputElement>;
      // email node should be placed before username
      expect(formChildNodes[0].name, 'email');
      expect(formChildNodes[1].name, 'username');
      expect(formChildNodes[2].type, 'password');
      expect(formChildNodes[3].type, 'submit');
```

**Review by mdebbar — APPROVED** [review]
LGTM

**Comment by JaredEzz** [comment]
Is there a good way to know when this would be rolled into stable or hotfixed? I tried to view on https://flutter-dashboard.appspot.com/#/build but it was a bit confusing

**Comment by willxie** [comment]
Thanks for the fix.

Has this been reverted? It works in beta channel for `3.12.0` but not the current latest `3.12.0-1.1.pre`

**Comment by JaredEzz** [comment]
Tested on stable 3.13 and bitwarden can autofill from the extension but no context menu shows up in the textfield

**Pull request merged**